README.md
*******************************************************

***********     DB Privilege Advisor     **************

*******************************************************
Author: SK Yang 

Website: https://skcave.wordpress.com/

This little program will help you to identify that your db settings have flaw or not,
then it will give you suggestion if there is a way to improve its security.

Requirement:
PHP 5+

Usage:
Checkout this project and fill your database connection info into config/config.php, then open index.php with your browser.

Disclaimer: 
I'm not responsible for any damage to your system while using it, so use it at your own risk.

Licensed under Apache 2.0, you may use this program freely for non-commericial use;
a license is required for commericial use.


changelog.txt
======== Change log v0.12 ========
- Change to MVC like file architrcture
- Check PostgreSQL privilege setting functions added
- Bug fixed

======== Change log v0.11 ========
- Check MySQL privilege setting functions added
- Bugs fixed

======== Change log v0.1 ========
- Check password strength
- Check if you're using this tool inside LAN or not
- Check port setting is default or not
- Check role that you're going to access

index.php
<?php
/*******************************************************
***********     DB Privilege Advisor     ***************
********************************************************
Author: SK Yang 
Website: https://skcave.wordpress.com/
File Version: 0.12

This little program will help you to identify that your db settings have flaw or not,
then it will give you suggestion if there is a way to improve its security.

Disclaimer: 
I'm not responsible for any damage to your system while using it, so use it at your own risk.

Licensed under Apache 2.0, you may use this program freely for non-commericial use;
a license is required for commericial use.
*/
error_reporting(E_ALL ^ E_NOTICE);

require_once(dirname(__FILE__) . '/func/func.php');

if($dbConf['type'] === 'mysql'){
	require_once(dirname(__FILE__) . '/func/mysql-func.php');
}
else if($dbConf['type'] === 'postgre'){
	require_once(dirname(__FILE__) . '/func/postgre-func.php');
}
?>
<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>DB Privilege Advisor</title>
</head>
<body>
	<h1>DB Privilege Advisor</h1><br/>
	<p>This program will help you to identify how secure your database is with current settings,<br/>
	then it will give you suggestion if there are anything you can do to improve its security.
	</p><br/>
	<table id='conn_info' width='400'>
		<tr>
			<center>Database Connection Infomation</center>
		</tr>
		<tr>
			<td width='25%'>Server type</td>
			<td width='25%'><?php echo $dbConf['type']; ?></td>
			<td></td>
		</tr>
		<tr>
			<td width='25%'>Server IP</td>
			<td width='25%'><?php echo $dbConf['address']; ?></td>
			<td><?php echo connIPChk(); ?></td>
		</tr>
		<tr>
			<td width='25%'>Port</td>
			<td width='25%'><?php echo $dbConf['port']; ?></td>
			<td><?php echo portChk(); ?></td>
		</tr>
		<tr>
			<td width='25%'>DB name</td>
			<td width='25%'><?php echo $dbConf['dbName']; ?></td>
			<td></td>
		</tr>
		<tr>
			<td width='25%'>User account</td>
			<td width='25%'><?php echo $dbConf['user']; ?></td>
			<td></td>
		</tr>
		<tr>
			<td width='25%'>Password</td>
			<td width='25%'><?php echo $dbConf['password']; ?></td>
			<td><?php echo pwdChk(); ?></td>
		</tr>
		<tr>
			<td width='25%'>Account role</td>
			<td width='25%'><?php echo $dbConf['rold']; ?></td>
			<td><?php echo roleChk(); ?></td>
		</tr>
	</table><br/>
	<?php 
		if($dbConf['type'] === 'mysql'){
			echo mysqlResult();
		}
		else if($dbConf['type'] === 'postgre'){
			echo postgreResult();
		}
	?>
	<br/>
</body>
</html>

func/func.php
<?php
/*******************************************************
***********     DB Privilege Advisor     ***************
********************************************************
Author: SK Yang 
Website: https://skcave.wordpress.com/
Version: 0.11

This file contains common functions for DB connection info examine.
*/
require_once(dirname(dirname(__FILE__)) . '/config/config.php');

// Grade your password strength
function pwdChk(){
	global $dbConf;
	$grade = 0;
	$str = '';
	$patterns = array('#[a-z]#','#[A-Z]#','#[0-9]#','/[?!"¢G$%^&*()`{}\[\]:@~;\'#<>?,.\/\\-=_+\|]/'); 
	$length = strlen($dbConf['password']);
	
	// More than 7 digit is considered as a line to cross weak grade	
	if($length >= 7){	
		if($length >= 10) $grade+=2;
		
		foreach($patterns as $pattern)
		{
			if(preg_match($pattern, $dbConf['password'], $matches))
			{
				$grade++;
			}
		} 
	}
	
	if($grade > 4) 		$str = '<font color="green">STRONG</font>';
	else if($grade > 2) $str = '<font color="#FFAA33">FAIR</font>';
	else $str = '<font color="red">WEAK</font>';	
	
	return $str;
}

// Check the IP that you given is inside the LAN or not
function connIPChk($ip = NULL){	
	if(!isset($ip)){
		global $dbConf;
		$ip = $dbConf['address'];
	}

	// Don't ask me why I don't check out if the ip is less or equal 255:255:255:255. 
	// It is a basic knowledge if you are going to use this tool!
	if( ($ip === 'localhost') ||
		($ip === '127.0.0.1') || 
		($ip === '::1') || 
		(preg_match('/10\./', $ip)) || 
		(preg_match('/172\.(1[6-9]|2[0-9]|3[0-1])\./', $ip)) || 
		(preg_match('/192\.168\./', $ip)))
	{
		return '<font color="green">You\'re using this tool inside the private network.</font>';
	}
	else
	{
		return '<font color="red">You\'re using this tool to test DB via public area network!'
		.' YOUR CONNECTION INFO MIGHT LEAK!</font>';
	}
}

// Check you port setting is the default or not
function portChk(){
	global $dbConf;
	
	if(($dbConf['type'] == 'mysql') && ($dbConf['port'] == '3306')){
		if( preg_match('/public/', connIPChk()) ){
			return '<font color="red">Your DB might suffer direct attack like brute force or DoS with this setting</font>';
		}
		else {
			return '<font color="#FFAA33">Default port is ok if you are in the LAN, but changing port is recommended</font>';
		}
	}
	
	return '<font color="green">Look good</font>';
}

// Check the role that you mentioned to use
function roleChk(){
	global $dbConf;
	$str = '';
	
	// Access from public
	if( preg_match('/public/', connIPChk()) ){
		if(($dbConf['role'] != 'app') || ($dbConf['role'] != 'unknown')){
			$str = '<font color="red">This account should not be used from public network!</font>';
		}
		else $str = '<font color="orange">Not recommend if there is no special purpose to accessing from public network</font>';
	}
	else{
		if(($dbConf['role'] != 'app')){
			$str = '<font color="orange">Use with caution</font>';
		}
		else $str = '<font color="green">Look good</font>';
	}
	
	return $str;
}



?>


config/config.php
<?php
/***************************************************
*************** DB Connection Config ***************
****************************************************
Fill your database connection info into those fields.
*/

$dbConf['user']		= 'test';
$dbConf['password'] = 'test';
$dbConf['dbName'] 	= 'test'; // Database Name
$dbConf['address'] 	= 'localhost'; // Server IP
$dbConf['port']		= '5432'; // Server Port

// The group name that this account member of(for Postgresql)
$dbConf['group']	= 'reader';

/*** DB Type ***
mysql: the test subject is MySQL.
postgre: the test subject is PostgreSQL.

(support mysql 5+, postgreSQL 9, didn't test the other version.)
*/
$dbConf['type']		= 'postgre'; 


/*** DB Role ***
super: super user of DB server.
replic: DB replication user.
project: individual DB project staff.
app: application user.
unknown: this program will just list the privliege it have. 
*/ 
$dbConf['role']		= 'unknown';

?>

func/postgre-func.php
<?php
/*******************************************************
***********     DB Privilege Advisor     ***************
********************************************************
Author: SK Yang 
Website: https://skcave.wordpress.com/
File Version: 0.1
Date: 05/05/2015

This file contained functions to examine the PostgreSQL settings.
*/
require_once(dirname(dirname(__FILE__)) . '/config/config.php');

// Determine that this user have the DDL or superuser privilege or not
function chkPriv($str){
	$retStr = 'Global ';
	
	if($str['rolsuper'] == true){
		$retStr .= 'SUPERUSER, ';
	}
	
	if($str['rolcreaterole'] == true){
		$retStr .= 'CREATE USER, ';
	}
	
	if($str['rolcreatedb'] == true){
		$retStr .= 'CREATE DATABASE, ';
	}
	
	if($str['rolcatupdate'] == true){
		$retStr .= 'CATALOGS MODIFY, ';
	}
	
	if($str['rolreplication'] == true){
		$retStr .= 'REPLICATION, ';
	}
	
	if($retStr == ''){
		$retStr .= 'no privilege';
	}
	
	return $retStr;
}

// Generate advice report for account current privileges
function getAdvice($role, $grantRep){
	$advice = '';

	if(preg_match('#Global#', $grantRep, $matches)){
		// Global superuser advice
		if(($role == 'super') && (preg_match('#SUPERUSER#', $grantRep, $matches))){
			$advice = '<font color="green">Global privilege settings look good.</font>';
		}
		// Global replication user advice
		else if(($role == 'replic') && (preg_match('#REPLICATION#', $grantRep, $matches))){
			$advice = '<font color="green">Global privilege settings look good.</font>';
		}
		// Other level user is not supposed to have any privilege on global
		else if(!preg_match('#no privilege#', $grantRep, $matches)){
			$advice = '<font color="red">User which is not a Admin or for Replication purposes are not suppose to have any global privilege.</font>';
		}
		// No privilege at global
		else{
			if($role == 'super'){
				$advice = '<font color="#FFAA33">Looks like its privilege is not set properly.</font> Grant all priveileges on global to this account.';				
			}
			else if($role == 'replic'){
				$advice = '<font color="#FFAA33">Looks like its privilege is not set properly.</font> Grant replication and CRUD privileges on global to this account.';
			}
			else $advice = '<font color="green">Global privilege settings look good.</font>';
		}
	}
	else {
		// Local development staff, got partial or full control of own database
		if(($role == 'project') && (preg_match('#SUPERUSER#', $grantRep, $matches) || preg_match('#development staff#', $grantRep, $matches))){
			$advice = '<font color="green">Local privilege settings look good.</font>';
		}
		// Local application user
		else if(($role == 'app') && (preg_match('#application user#', $grantRep, $matches))){
			$advice = '<font color="green">Local privilege settings look good.</font>';
		}
		else if(!preg_match('#no privilege#', $grantRep, $matches)){
			$advice = '<font color="#FFAA33">Looks like this account do not have any privilege to this database.</font>';
		}
		else $advice = '<font color="green">Local privilege settings look good.</font>';
	}
	
	return $advice;
}

// Perform test and generate result form
function postgreResult(){
	global $dbConf;
	$report = '<table id="postgre_report" width="600">';
	
	try{
		$conn = new PDO('pgsql:host=' . $dbConf['address'] . ';port=' . $dbConf['port'] . ';user=' . $dbConf['user'] . ';password="' . $dbConf['password'] .';dbname=' . $dbConf['dbName']);
		
		$rs = $conn->query('SELECT rolsuper, rolcreaterole, rolcreatedb, rolcatupdate, rolreplication FROM pg_roles WHERE rolname="' . $dbConf['user'] . '"');
		
		// Fetch DDL grant results
		$row = $rs->fetch();
		$grantRep = chkPriv($row);
		$report .= '<tr><td>Database Privilege</td><td>';
			
		$advice = getAdvice($dbConf['role'], $grantRep);
		
		$report .= $grantRep . '</td><td>' . $advice . '</td></tr>';
		
		$conn = null;
		$report .= '</table>';
		return $report;
	}
	catch(PDOException $e) {
		$str  = '<tr><td>Error: ' . $e->getMessage() . '</td></tr></table>';
		$conn = null;
		return $str;
	}
}

?>

func/mysql-func.php
<?php
/*******************************************************
***********     DB Privilege Advisor     ***************
********************************************************
Author: SK Yang 
Website: https://skcave.wordpress.com/
Version: 0.11

This file contained functions to examine the MySQL settings.
*/
require_once(dirname(dirname(__FILE__)) . '/config/config.php');


// Determine which role that this account actually is
function chkGrant($str = NULL){
	$retStr = '';
	$patterns = array('#SELECT#', '#INSERT#', '#UPDATE#', '#DELETE#', '#FILE#', '#EXECUTE#', '#LOCK TABLES#',
				'#CREATE#', '#ALTER#', '#INDEX#', '#DROP#', '#EVENT#', '#CREATE VIEW#', '#CREATE TEMPORARY TABLES#', '#SHOW VIEW#', '#CREATE ROUTINE#', '#ALTER ROUTINE#', '#TRIGGER#', 
				'#REPLICATION CLIENT#', '#REPLICATION SLAVE#',
				'#RELOAD#', '#PROCESS#', '#SHUTDOWN#', '#SHOW DATABASES#', '#CREATE USER#'); 
	
	if(!isset($str)){
		return 'Error occured: No data.';
	}
	else{
		// Check if this is the message of global privilege
		if(preg_match('/\*\.\*/', $str, $matches)){
			$retStr = 'Global ';
		} 
		else $retStr = 'Local ';
		
		// Determine user privilege level
		if(preg_match('#ALL PRIVILEGES#', $str, $matches) || preg_match('#SUPER#', $str, $matches)){
			$retStr .= 'SUPERUSER';
		}
		else if(preg_match('#USAGE#', $str, $matches)){
			$retStr .= 'no privilege';
		}
		else {
			$lvl = 0;
			for($i = 0; $i < 25; $i++){
				if(preg_match($patterns[$i], $str, $matches)){
					$lvl = $i;
				}
			}
			
			if($lvl < 7){ // application user level
				$retStr .= 'application user';
			}
			else if($lvl < 18){// development staff level
				$retStr .= 'development staff';
			}
			else if($lvl < 20){// replication user level
				$retStr .= 'replication user';
			}
			else{// server admin level
				$retStr .= 'SUPERUSER';
			}
		}
		
	}
	
	return $retStr;
}
// Generate advice report for account current privileges
function getAdvice($role, $grantRep){
	$advice = '';

	if(preg_match('#Global#', $grantRep, $matches)){
		// Global superuser advice
		if(($role == 'super') && (preg_match('#SUPERUSER#', $grantRep, $matches))){
			$advice = '<font color="green">Global privilege settings look good.</font>';
		}
		// Global replication user advice
		else if(($role == 'replic') && (preg_match('#REPLICATION#', $grantRep, $matches))){
			$advice = '<font color="green">Global privilege settings look good.</font>';
		}
		// Other level user is not supposed to have any privilege on global
		else if(!preg_match('#no privilege#', $grantRep, $matches)){
			$advice = '<font color="red">User which is not a Admin or for Replication purposes are not suppose to have any global privilege.</font>';
		}
		// No privilege at global
		else{
			if($role == 'super'){
				$advice = '<font color="#FFAA33">Looks like its privilege is not set properly.</font> Grant all priveileges on global to this account.';				
			}
			else if($role == 'replic'){
				$advice = '<font color="#FFAA33">Looks like its privilege is not set properly.</font> Grant replication and CRUD privileges on global to this account.';
			}
			else $advice = '<font color="green">Global privilege settings look good.</font>';
		}
	}
	else {
		// Local development staff, got partial or full control of own database
		if(($role == 'project') && ((preg_match('#SUPERUSER#', $grantRep, $matches)) || (preg_match('#development staff#', $grantRep, $matches)))){
			$advice = '<font color="green">Local privilege settings look good.</font>';
		}
		// Local application user
		else if(($role == 'app') && (preg_match('#application user#', $grantRep, $matches))){
			$advice = '<font color="green">Local privilege settings look good.</font>';
		}
		else if(preg_match('#no privilege#', $grantRep, $matches)){
			$advice = '<font color="#FFAA33">Looks like this account do not have any privilege to this database.</font>';
		}
		else $advice = '<font color="green">Local privilege settings look good.</font>';
	}
	
	return $advice;
}

// Perform test and generate result form
function mysqlResult(){
	global $dbConf;
	$report = '<table id="mysql_report" width="600"><tr><center>User Privilege Report</center></tr>';
	
	try{
		$conn = new PDO('mysql:host=' . $dbConf['address'] . ';port=' . $dbConf['port'] . ';charset=utf8', $dbConf['user'], $dbConf['password']);
		$rs = $conn->query('SHOW GRANTS');
		
		// Fetch grant results
		while($row = $rs->fetch()){
			$report .= '<tr><td>Database Privilege</td><td>' . $row[0] . '</td><td>';
			
			$grantRep = chkGrant($row[0]);
			$advice = getAdvice($dbConf['role'], $grantRep);
			
			$report .= $grantRep . '</td><td>' . $advice . '</td></tr>';
		}
		
		$conn = null;
		$report .= '</table>';
		return $report;
	}
	catch(PDOException $e) {
		$str  = '<tr><td>Error: ' . $e->getMessage() . '</td></tr></table>';
		$conn = null;
		return $str;
	}
	
	
}
?>